log-manager/src/main/java/io/airlift/log/RollingFileHandler.java

```diff
@@ -1,89 +1,231 @@
+    private LogFileName currentOutputFileName;
+    @GuardedBy("this")
+    private OutputStream currentOutputStream;
+    @GuardedBy("this")
+    private long currentFileSize;
+    private final LogHistoryManager historyManager;
+
+    private final BlockingQueue<byte[]> queue = new ArrayBlockingQueue<>(MAX_BATCH_COUNT);
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final Thread thread;
 
-    private final AsyncAppenderBase<String> asyncAppender;
+    private final ExecutorService compressionExecutor;
 
-    public RollingFileHandler(String filename, int maxHistory, long maxSizeInBytes)
+    public static RollingFileHandler createRollingFileHandler(String filename, DataSize maxFileSize, DataSize maxTotalSize, CompressionType compressionType)
     {
-        setFormatter(new StaticFormatter());
+        RollingFileHandler handler = new RollingFileHandler(filename, maxFileSize, maxTotalSize, compressionType);
+        handler.start();
+        return handler;
+    }
 
-        ContextBase context = new ContextBase();
+    private RollingFileHandler(String filename, DataSize maxFileSize, DataSize maxTotalSize, CompressionType compressionType)
+    {
+        requireNonNull(filename, "filename is null");
+        requireNonNull(maxFileSize, "maxFileSize is null");
+        requireNonNull(maxTotalSize, "maxTotalSize is null");
+        requireNonNull(compressionType, "compressionType is null");
 
-        recoverTempFiles(filename);
+        this.maxFileSize = maxFileSize.toBytes();
+        this.compressionType = compressionType;
 
-        RollingFileAppender<String> fileAppender = new RollingFileAppender<>();
-        TimeBasedRollingPolicy<String> rollingPolicy = new TimeBasedRollingPolicy<>();
-        SizeAndTimeBasedFNATP<String> triggeringPolicy = new SizeAndTimeBasedFNATP<>();
+        link = Paths.get(filename);
 
-        rollingPolicy.setContext(context);
-        rollingPolicy.setFileNamePattern(filename + "-%d{yyyy-MM-dd}.%i.log.gz");
-        rollingPolicy.setMaxHistory(maxHistory);
-        rollingPolicy.setTimeBasedFileNamingAndTriggeringPolicy(triggeringPolicy);
-        rollingPolicy.setParent(fileAppender);
+        thread = new Thread(this::logging);
+        thread.setName("log-writer");
+        thread.setDaemon(true);
+
+        setFormatter(new StaticFormatter());
+
+        // ensure log directory can be created
+        try {
+            MoreFiles.createParentDirectories(link);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
 
-        // Limit total log files occupancy on disk. Ideally we would keep exactly
-        // `maxHistory` files (not logging periods). This is closest currently possible.
-        rollingPolicy.setTotalSizeCap(new FileSize(LongMath.saturatedMultiply(maxSizeInBytes, maxHistory)));
+        // convert from legacy logger
+        try {
+            LegacyRollingFileHandler.recoverTempFiles(filename);
```

Comment:
do we need this?

Reply:
We support upgrading from the old code, and the old code may have left over temp files that have to be cleaned up.